funsor/cnf.py

```diff
@@ -231,14 +231,14 @@ def eager_contraction_to_binary(red_op, bin_op, reduced_vars, lhs, rhs):
 def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
     if not all(term.dtype == "real" for term in terms):
         raise NotImplementedError('TODO')
-    return _eager_contract_tensors(reduced_vars, terms, backend="torch")
+    return _eager_contract_tensors(reduced_vars, terms, backend=terms[0].backend)
 
 
 @eager.register(Contraction, ops.LogAddExpOp, ops.AddOp, frozenset, Tensor, Tensor)
 def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
     if not all(term.dtype == "real" for term in terms):
         raise NotImplementedError('TODO')
-    return _eager_contract_tensors(reduced_vars, terms, backend="pyro.ops.einsum.torch_log")
+    return _eager_contract_tensors(reduced_vars, terms, backend=terms[0].logsumexp_backend)
```

Comment:
Tensor.logsumexp_backend is either `pyro.ops.einsum.torch_log` or `funsor.numpy`